internal/vcs/git/diff_search.go

```
@@ -492,3 +483,16 @@ func filterAndResolveRefs(ctx context.Context, repo gitserver.Repo, refs []strin
 	}
 	return filtered, nil
 }
+
+// withDeadlinePercentage returns a context which expires once p of the
+// remaining time has passed. p decimal fraction in [0,1]. For example to
+// expire in half the remaining time set p to 0.5. To expire in 80% of the
+// remaining time, set p to 0.8.
+func withDeadlinePercentage(ctx context.Context, p float64) (context.Context, context.CancelFunc) {
+	if deadline, ok := ctx.Deadline(); ok {
+		now := time.Now()
+		d := time.Duration(float64(deadline.Sub(now)) * p)
+		return context.WithDeadline(ctx, now.Add(d))
+	}
+	return ctx, func() {}
```

This could be misleading, couldn't it? The caller could expect the cancel function to work. Should we return an error instead if no deadline is set?

---

you are right! This is actually a bug in I inherited from the old `withTimeout` helper I am replacing. In practice we always have a timeout, but I will fix :)